xos.h

```diff
@@ -7,77 +7,130 @@
 #include <string>
 #include <vector>
 
-namespace xos {
-
-class Token {
- public:
-  enum Kind {
-    eof = -1,
-    main = 1,
-    out = 2,
-    lparen = 3,
-    rparen = 4,
-    arrow = 5,
-    string = 6,
-    identifier = 7,
-    number = 8,
+namespace xos
+{
+
+  class Token
+  {
+  public:
+    enum Kind
+    {
+      eof = -1,
+      main = 1,
+      out = 2,
+      lparen = 3,
+      rparen = 4,
+      arrow = 5,
+      tstring = 6,
+      identifier = 7,
+      number = 8,
+    };
+
+    /** The token can be a character, string, a number, or anything else.
+     * -- Switch to using variant instead of union?
+     */
+    union val
+    {
+      char cval;
+      const char *sval;
+      int inum;
+      bool eof;
+      void *any;
+    } v;
+
+    Token(Kind kind, uint32_t row, uint32_t col, val v)
+        : kind_(kind), row_(row), col_(col), v_(v) {}
+    Token() = default;
+
+    Kind getKind() const { return kind_; }
+    uint32_t getRow() const { return row_; }
+    uint32_t getCol() const { return col_; }
+    val getVal() const { return v_; }
+
+  private:
+    Kind kind_;
+    uint32_t row_, col_;
+    val v_;
```

Comment:
Rather than having a val that has the actual value of the token we parsed, this should just be a string containing the actual characters of the thing we parsed. Anything relating to knowing the "value" of the underlying token should be done in the parser via a node. We'll still need this logic for then, but the lexer shouldn't have to do anything regarding knowing the "value" for a token. (To get an int, this would imply that the lexer has to "parse" an int from a string.)